kv/kvserver/storeliveness: TestStoreLivenessDiskStall failed

kv/kvserver/storeliveness.TestStoreLivenessDiskStall [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18263807?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18263807?buildTab=artifacts#/) on master @ [0806ee44e36b62dce75175702b5bfb3db03e0577](https://github.com/cockroachdb/cockroach/commits/0806ee44e36b62dce75175702b5bfb3db03e0577):

Fatal error:

```
panic: test timed out after 5m0s
running tests:
	TestStoreLivenessDiskStall (4m40s)

```
Stack: 

```
goroutine 30599 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* github.com/cockroachdb/pebble/record.NewLogWriter.func2()
* 	external/com_github_cockroachdb_pebble/record/log_writer.go:564 +0x60
* created by github.com/cockroachdb/pebble/record.NewLogWriter in goroutine 16907
* 	external/com_github_cockroachdb_pebble/record/log_writer.go:563 +0x44c
* 
* goroutine 17059 [sync.Cond.Wait, 2 minutes]:
* sync.runtime_notifyListWait(0x400262c0a8, 0x1b1)
* 	GOROOT/src/runtime/sema.go:569 +0x154
* sync.(*Cond).Wait(0x400262c098)
* 	GOROOT/src/sync/cond.go:70 +0xcc
* github.com/cockroachdb/pebble/record.(*flusherCond).Wait(...)
* 	external/com_github_cockroachdb_pebble/record/log_writer.go:371
* github.com/cockroachdb/pebble/record.(*LogWriter).flushLoop(0x400262c000, {0x478de00?, 0x9bce420?})
* 	external/com_github_cockroachdb_pebble/record/log_writer.go:645 +0x670
* runtime/pprof.Do({0x6517aa8?, 0x9bce420?}, {{0x40007f41a0?, 0x48ca48?, 0x0?}}, 0x4004ae2fb8)
* 	GOROOT/src/runtime/pprof/runtime.go:51 +0x80
* github.com/cockroachdb/pebble/record.NewLogWriter.func2()
* 	external/com_github_cockroachdb_pebble/record/log_writer.go:564 +0x60
* created by github.com/cockroachdb/pebble/record.NewLogWriter in goroutine 16907
* 	external/com_github_cockroachdb_pebble/record/log_writer.go:563 +0x44c
* 
* goroutine 17058 [select]:
* github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker.func1()
* 	external/com_github_cockroachdb_pebble/vfs/disk_health.go:254 +0x9c
* created by github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker in goroutine 16907
* 	external/com_github_cockroachdb_pebble/vfs/disk_health.go:249 +0x64
* 
* goroutine 27632 [sync.Cond.Wait, 2 minutes]:
* sync.runtime_notifyListWait(0x40104c40a8, 0x25a)
* 	GOROOT/src/runtime/sema.go:569 +0x154
* sync.(*Cond).Wait(0x40104c4098)
* 	GOROOT/src/sync/cond.go:70 +0xcc
* github.com/cockroachdb/pebble/record.(*flusherCond).Wait(...)
* 	external/com_github_cockroachdb_pebble/record/log_writer.go:371
* github.com/cockroachdb/pebble/record.(*LogWriter).flushLoop(0x40104c4000, {0x478de00?, 0x9bce420?})
* 	external/com_github_cockroachdb_pebble/record/log_writer.go:645 +0x670
* runtime/pprof.Do({0x6517aa8?, 0x9bce420?}, {{0x40007f41a0?, 0x0?, 0x0?}}, 0x400c9cb7b8)
* 	GOROOT/src/runtime/pprof/runtime.go:51 +0x80
* github.com/cockroachdb/pebble/record.NewLogWriter.func2()
* 	external/com_github_cockroachdb_pebble/record/log_writer.go:564 +0x60
* created by github.com/cockroachdb/pebble/record.NewLogWriter in goroutine 18035
* 	external/com_github_cockroachdb_pebble/record/log_writer.go:563 +0x44c
* 
* goroutine 27631 [select]:
* github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker.func1()
* 	external/com_github_cockroachdb_pebble/vfs/disk_health.go:254 +0x9c
* created by github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFile).startTicker in goroutine 18035
* 	external/com_github_cockroachdb_pebble/vfs/disk_health.go:249 +0x64
* 
*
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreLivenessDiskStall.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-45780

--- a/pkg/kv/kvserver/storeliveness/multi_store_test.go
+++ b/pkg/kv/kvserver/storeliveness/multi_store_test.go
@@ -278,7 +278,6 @@ func TestStoreLivenessDiskStall(t *testing.T) {
 	checkSupport(t, tc, checkSupportFn)
 
 	// Stop blocking writes.
-	testEngine.SignalToUnblock()
 	testEngine.SetBlockOnWrite(false)
 
 	// Ensure all-to-all support again.

--- a/pkg/kv/kvserver/storeliveness/store_liveness_test.go
+++ b/pkg/kv/kvserver/storeliveness/store_liveness_test.go
@@ -100,7 +100,7 @@ func TestStoreLiveness(t *testing.T) {
 					case "error-on-write":
 						var errorOnWrite bool
 						d.ScanArgs(t, "on", &errorOnWrite)
-						engine.errorOnWrite = errorOnWrite
+						engine.SetErrorOnWrite(errorOnWrite)
 						return ""
 
 					case "debug-requester-state":

--- a/pkg/kv/kvserver/storeliveness/support_manager_test.go
+++ b/pkg/kv/kvserver/storeliveness/support_manager_test.go
@@ -348,7 +348,6 @@ func TestSupportManagerDiskStall(t *testing.T) {
 	require.True(t, supported)
 
 	// Stop blocking writes.
-	engine.SignalToUnblock()
 	engine.SetBlockOnWrite(false)
 
 	// Ensure the heartbeat is unblocked and sent out.

--- a/pkg/kv/kvserver/storeliveness/testutils.go
+++ b/pkg/kv/kvserver/storeliveness/testutils.go
@@ -7,6 +7,7 @@ package storeliveness
 
 import (
 	"context"
+	"sync"
 
 	slpb "github.com/cockroachdb/cockroach/pkg/kv/kvserver/storeliveness/storelivenesspb"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
@@ -21,27 +22,24 @@ type TestEngine struct {
 	storage.Engine
 	storeID      slpb.StoreIdent
 	mu           syncutil.Mutex
-	blockingCh   chan struct{}
+	cond         sync.Cond
 	blockOnWrite bool
 	errorOnWrite bool
 }
 
 func NewTestEngine(storeID slpb.StoreIdent) *TestEngine {
-	return &TestEngine{
-		Engine:     storage.NewDefaultInMemForTesting(),
-		storeID:    storeID,
-		blockingCh: make(chan struct{}, 1),
+	te := &TestEngine{
+		Engine:  storage.NewDefaultInMemForTesting(),
+		storeID: storeID,
 	}
+	te.cond.L = &te.mu
+	return te
 }
 
 func (te *TestEngine) NewBatch() storage.Batch {
-	te.mu.Lock()
-	defer te.mu.Unlock()
 	return TestBatch{
-		Batch:        te.Engine.NewBatch(),
-		blockingCh:   te.blockingCh,
-		blockOnWrite: te.blockOnWrite,
-		errorOnWrite: te.errorOnWrite,
+		Batch: te.Engine.NewBatch(),
+		te:    te,
 	}
 }
 
@@ -49,37 +47,36 @@ func (te *TestEngine) SetBlockOnWrite(bow bool) {
 	te.mu.Lock()
 	defer te.mu.Unlock()
 	te.blockOnWrite = bow
+	te.cond.Broadcast()
 }
 
-func (te *TestEngine) SignalToUnblock() {
-	te.blockingCh <- struct{}{}
+func (te *TestEngine) blockOrErrorOnWrite() error {
+	te.mu.Lock()
+	defer te.mu.Unlock()
+	for te.blockOnWrite {
+		te.cond.Wait()
+	}
+	if te.errorOnWrite {
+		return errors.New("error on write")
+	}
+	return nil
 }
 
 func (te *TestEngine) PutUnversioned(key roachpb.Key, value []byte) error {
-	te.mu.Lock()
-	defer te.mu.Unlock()
-	if te.blockOnWrite {
-		<-te.blockingCh
-	}
-	if te.errorOnWrite {
-		return errors.New("error writing")
+	if err := te.blockOrErrorOnWrite(); err != nil {
+		return err
 	}
 	return te.Engine.PutUnversioned(key, value)
 }
 
 type TestBatch struct {
 	storage.Batch
-	blockingCh   chan struct{}
-	blockOnWrite bool
-	errorOnWrite bool
+	te *TestEngine
 }
 
 func (tb TestBatch) Commit(sync bool) error {
-	if tb.blockOnWrite {
-		<-tb.blockingCh
-	}
-	if tb.errorOnWrite {
-		return errors.New("error committing batch")
+	if err := tb.te.blockOrErrorOnWrite(); err != nil {
+		return err
 	}
 	return tb.Batch.Commit(sync)
 }

--- a/pkg/kv/kvserver/storeliveness/testutils.go
+++ b/pkg/kv/kvserver/storeliveness/testutils.go
@@ -37,7 +37,7 @@ func NewTestEngine(storeID slpb.StoreIdent) *TestEngine {
 }
 
 func (te *TestEngine) NewBatch() storage.Batch {
-	return TestBatch{
+	return &TestBatch{
 		Batch: te.Engine.NewBatch(),
 		te:    te,
 	}
@@ -80,7 +80,7 @@ type TestBatch struct {
 	te *TestEngine
 }
 
-func (tb TestBatch) Commit(sync bool) error {
+func (tb *TestBatch) Commit(sync bool) error {
 	if err := tb.te.blockOrErrorOnWrite(); err != nil {
 		return err
 	}

--- a/pkg/kv/kvserver/storeliveness/testutils.go
+++ b/pkg/kv/kvserver/storeliveness/testutils.go
@@ -50,6 +50,12 @@ func (te *TestEngine) SetBlockOnWrite(bow bool) {
 	te.cond.Broadcast()
 }
 
+func (te *TestEngine) SetErrorOnWrite(eow bool) {
+	te.mu.Lock()
+	defer te.mu.Unlock()
+	te.errorOnWrite = eow
+}
+
 func (te *TestEngine) blockOrErrorOnWrite() error {
 	te.mu.Lock()
 	defer te.mu.Unlock()

